source/docs/software/basic-programming/functions-as-data.rst

```diff
@@ -0,0 +1,118 @@
+
+Treating Functions as Data in Java
+----------------------------------
+
+Java represents functions-as-data as instances of "functional interfaces" <TODO: link to docs>.  A functional interface is a special kind of class that has only a single method - since Java was originally designed strictly for object-oriented programming, it has no way of representing a single function detached from a class.  Instead, it defines a particular group of classes that *only* represent single functions.  Each type of function signature has its own functional interface, which is an interface with a single function definition of that signature.
+
+This might sound complicated, but in the context of WPILib we don't really need to worry much about using the functional interfaces themselves - the code that does that is internal to WPILib.  Instead, all we need to know is how to pass a function that we've written to a method that takes a functional interface as a parameter.  For a simple example, consider the signature of ``Commands.runOnce`` (which creates an ``InstantCommand`` that, when scheduled, runs the given function once and then terminates):
+
+.. note:: The ``requirements`` parameter is explained in the Command-based documentation, and will not be discussed here <TODO: link>.
+.. code-block:: java
+
+   public static CommandBase runOnce(Runnable action, Subsystem... requirements)
+
+``runOnce`` expects us to give it a ``Runnable`` parameter (named ``action``).  A ``Runnable`` is the Java term for a function that takes no parameters and returns no value.  When we call ``runOnce``, we need to give it a function with no parameters and no return value.  There are two ways to do this: we can refer to some existing function using a "method reference", or we can define the function we want inline using a "lambda expression".
+
+Method References
+^^^^^^^^^^^^^^^^^
+
+A method reference lets us pass an already-existing function as our ``Runnable``:
+
+.. code-block:: java
+
+   // Create an InstantCommand that runs the `resetEncoders` method of the `drivetrain` object
+   Command disableCommand = runOnce(drivetrain::resetEncoders, drivetrain);
+
+The expression ``drivetrain::resetEncoders`` is a reference to the ``resetEncoders`` method of the ``drivetrain`` object.  It is not a method *call* - this line of code does not *itself* reset the encoders of the drivetrain.  Instead, it returns a ``Command`` that will do so *when it is scheduled.*
+
+Remember that in order for this to work, ``resetEncoders`` must be a ``Runnable`` - that is, it must take no parameters and return no value.  So, its signature must look like this:
+
+.. code-block:: java
+
+   // void because it returns no parameters, and has an empty parameter list
+   public void resetEncoders()
+
+If the function signature does not match this, Java will not be able to interpret the method reference as a ``Runnable`` and the code will not compile.
```

Comment:
This is accurate-but-not-really. In this case, the method reference must be a `Runnable` because `runOnce` expects a `Runnable`.  Other common usecases like `Trigger` constructors require `BooleanSuppliers` so the method would have to be `public boolean getSpeed()` in order for `new Trigger(shooter::getSpeed)` to compile.

Reply:
The case given is that of a `Runnable` in particular, and the text says as much; other functional interfaces can be discovered by readers in time.